src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -51,51 +65,84 @@ public class AsyncSemaphore : IDisposable
+        public Task<Releaser> EnterAsync(CancellationToken cancellationToken = default) => this.EnterAsync(Timeout.InfiniteTimeSpan, cancellationToken);
 
         /// <summary>
         /// Requests access to the lock.
         /// </summary>
         /// <param name="timeout">A timeout for waiting for the lock.</param>
         /// <param name="cancellationToken">A token whose cancellation signals lost interest in the lock.</param>
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default(CancellationToken))
+        /// <returns>
+        /// A task whose result is a releaser that should be disposed to release the lock.
+        /// This task may be canceled if <paramref name="cancellationToken"/> is signaled or <paramref name="timeout"/> expires.
+        /// </returns>
+        /// <exception cref="OperationCanceledException">Thrown when <paramref name="cancellationToken"/> is canceled or the <paramref name="timeout"/> expires before semaphore access is granted.</exception>
+        /// <exception cref="ObjectDisposedException">Thrown when this semaphore is disposed before semaphore access is granted.</exception>
+        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default)
         {
-            if (this.disposed)
+            if (cancellationToken.IsCancellationRequested)
             {
-                return DisposedReleaserTask;
+                return Task.FromCanceled<Releaser>(cancellationToken);
             }
 
-            return this.LockWaitingHelper(this.semaphore.WaitAsync(timeout, cancellationToken));
+            lock (this.syncObject)
+            {
+                if (this.disposed)
+                {
+                    return DisposedReleaserTask;
+                }
+
+                if (this.CurrentCount > 0)
+                {
+                    this.CurrentCount--;
+                    return this.uncontestedReleaser;
+                }
+                else if (timeout == TimeSpan.Zero)
+                {
+                    return CanceledReleaser;
+                }
+                else
+                {
+                    var node = this.GetNode(new WaiterInfo(this, cancellationToken));
+
+                    // Careful: consider that if the token was cancelled just now (after we checked it on entry to this method)
+                    // or the timeout expires,
+                    // then this Register method may *inline* the handler we give it, reversing the apparent order of execution with respect to
+                    // the code that follows this Register call.
+                    node.Value.CancellationTokenRegistration = cancellationToken.Register(CancellationHandler, node.Value);
+                    if (timeout != Timeout.InfiniteTimeSpan)
```

Comment:
If cancellation of the token were requested here, would that leave the timer undisposed?

Reply:
Probably. I've made a change to fix that.